x-pack/plugins/rule_registry/server/rule_registry/rule_type_helpers/create_lifecycle_rule_type_factory.ts

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import * as t from 'io-ts';
+import { isLeft } from 'fp-ts/lib/Either';
+import v4 from 'uuid/v4';
+import { AlertInstance } from '../../../../alerting/server';
+import { ActionVariable, AlertInstanceState } from '../../../../alerting/common';
+import { RuleParams, RuleType } from '../../types';
+import { DefaultFieldMap } from '../defaults/field_map';
+import { OutputOfFieldMap } from '../field_map/runtime_type_from_fieldmap';
+import { PrepopulatedRuleEventFields } from '../create_scoped_rule_registry_client/types';
+import { RuleRegistry } from '..';
+
+type UserDefinedAlertFields<TFieldMap extends DefaultFieldMap> = Omit<
+  OutputOfFieldMap<TFieldMap>,
+  PrepopulatedRuleEventFields | 'kibana.rac.alert.id' | 'kibana.rac.alert.uuid' | '@timestamp'
+>;
+
+type LifecycleAlertService<
+  TFieldMap extends DefaultFieldMap,
+  TActionVariable extends ActionVariable
+> = (alert: {
+  id: string;
+  fields: UserDefinedAlertFields<TFieldMap>;
+}) => AlertInstance<AlertInstanceState, { [key in TActionVariable['name']]: any }, string>;
+
+type CreateLifecycleRuleType<TFieldMap extends DefaultFieldMap> = <
+  TRuleParams extends RuleParams,
+  TActionVariable extends ActionVariable
+>(
+  type: RuleType<
+    TFieldMap,
+    TRuleParams,
+    TActionVariable,
+    { alertWithLifecycle: LifecycleAlertService<TFieldMap, TActionVariable> }
+  >
+) => RuleType<TFieldMap, TRuleParams, TActionVariable>;
+
+const trackedAlertStateRt = t.type({
+  alertId: t.string,
+  alertUuid: t.string,
+  started: t.string,
+});
+
+const wrappedStateRt = t.type({
+  wrapped: t.record(t.string, t.unknown),
+  trackedAlerts: t.record(t.string, trackedAlertStateRt),
+});
+
+export function createLifecycleRuleTypeFactory<
+  TRuleRegistry extends RuleRegistry<DefaultFieldMap>
+>(): TRuleRegistry extends RuleRegistry<infer TFieldMap>
+  ? CreateLifecycleRuleType<TFieldMap>
+  : never;
+
+export function createLifecycleRuleTypeFactory(): CreateLifecycleRuleType<DefaultFieldMap> {
```

Comment:
This is a great abstraction here @dgieselaar -- love it! 🙂 🚀 

In chatting with @madirey we're gonna start creating a few different rule type factories to retrofit the existing `security-solution` rule types and also for testing out other facets for finalizing the root schema, then can review all the different ones we have and see if there's any opportunity for consolidation. 

Curious how we can attach common functionality between `services`, whether we chain factories, or just have a little duplication and each factory leverages common/sharable pure functions. Example here being exceptions logic, or de-duplication. We can sort this out as we add more factories/reference rules though 🙂

Reply:
We _should_ be able to chain factories. Currently it's injecting new services but maybe currying or hoc will make more sense, I need to investigate.